jvm/core/src/main/java/ml/dmlc/tvm/rpc/ConnectTrackerServerProcessor.java

```diff
@@ -0,0 +1,227 @@
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ml.dmlc.tvm.rpc;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.ServerSocket;
+import java.net.Socket;
+import java.net.BindException;
+import java.net.InetSocketAddress;
+import java.net.SocketAddress;
+import java.net.SocketException;
+import java.net.SocketTimeoutException;
+
+
+/**
+ * Server processor with tracker connection (based on standalone).
+ */
+public class ConnectTrackerServerProcessor implements ServerProcessor {
+  private ServerSocket server;
+  private final SocketFileDescriptorGetter socketFileDescriptorGetter;
+  private final String trackerHost;
+  private final int trackerPort;
+  private final String key;
+  private final String matchKey;
+  private int serverPort = 5001;
+  public static final int MAX_SERVER_PORT = 5555;
+  // time to wait before aborting tracker connection (ms)
+  public static final int TRACKER_TIMEOUT = 6000;
+  // time to wait for a connection before refreshing tracker connection (ms)
+  public static final int STALE_TRACKER_TIMEOUT = 300000;
+  // time to wait if no timeout value is specified (seconds)
+  public static final int HARD_TIMEOUT_DEFAULT = 300;
+  private RPCWatchdog watchdog;
+
+  public ConnectTrackerServerProcessor(String trackerHost, int trackerPort, String key,
+      SocketFileDescriptorGetter sockFdGetter, RPCWatchdog watchdog) throws IOException {
+    while (true) {
+        try {
+            this.server = new ServerSocket(serverPort);
+            server.setSoTimeout(STALE_TRACKER_TIMEOUT);
+            break;
+        } catch (BindException e) {
+            System.err.println(serverPort);
+            System.err.println(e);
+            serverPort++;
+            if (serverPort > MAX_SERVER_PORT) {
+                throw e;
+            }
+        }
+    }
+    System.err.println("using port: " + serverPort);
+    this.socketFileDescriptorGetter = sockFdGetter;
+    this.trackerHost = trackerHost;
+    this.trackerPort = trackerPort;
+    this.key = key;
+    this.matchKey = key + ":" + Math.random();
+    this.watchdog = watchdog;
+  }
+
+  public String getMatchKey() {
+    return matchKey;
+  }
+
+  @Override public void terminate() {
+    try {
+      server.close();
+    } catch (IOException e) {
+      e.printStackTrace();
+    }
+  }
+
+  @Override public void run() {
+    Socket trackerSocket = null;
+    String recvKey = null;
+    try {
+      trackerSocket = connectToTracker();
```

Comment:
Consider wrap a standalone class TrackerClient and move all related logics to there, to make the logic cleaner